runtime/caml/config.h

```diff
@@ -266,4 +266,15 @@ typedef uint64_t uintnat;
 /* Default size of runtime_events ringbuffers, in words, in powers of two */
 #define Default_runtime_events_log_wsize 16
 
+/* Assumed size of cache line. This value can be bigger than the actual L1
+   cache line size. Atomics allocated with aligned constructor are
+   memory-aligned this value to avoid false sharing of cache line. */
+#ifdef TARGET_s390x
+   #define Cache_line_bsize 256
+#else
+   /* Many 64-bit processors have 128-byte cache line. Even with 64-byte cache
+   line, it may be beneficial to have twice that. */
```

Comment:
I bite: why is it beneficial to use 2 cache lines where one would suffice (intuitively)?

Reply:
L2 spatial prefetcher on Intel 64 prefetches 128-byte chunks. If two domains act on atomics residing in the same chunk and prefetcher kicks in, one will steal others cache line. 
* https://docs.rs/crossbeam/latest/crossbeam/utils/struct.CachePadded.html#size-and-alignment 
* https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding

Sadly, I don't have any high-quality references. I'm happy to spend a little time trying to prove that it matters for us, and if there's no effect, make it 64-byte on x86 64 (and RISC-V, I think).

Reply:
I wondered the same. The best reference on stackoverflow seems to be https://stackoverflow.com/questions/72126606/should-the-cache-padding-size-of-x86-64-be-128-bytes and reading it in details shows that it is very debated whether the spatial prefetcher influences destructive interference.

For the record I ran the test found there on my machine (AMD Zen 3) and I do not see benefits of choosing 128 over 64. In addition, gcc decided that [std::hardware_constructive_interference_size and std::hardware_destructive_interference_size](https://en.cppreference.com/w/cpp/thread/hardware_destructive_interference_size) should both be 64 on my platform.

If the right value is machine-dependent for the same compilation target, could it be implemented as something else than a compile-time constant?

Reply:
One hypothesis from https://stackoverflow.com/a/72127222 is that without actual false sharing (within the same cache line), there is not effect of the L2 spatial prefetcher (each thread ending up exclusive owner of its cache line). One hypothesis from https://www.mail-archive.com/gcc-patches@gcc.gnu.org/msg265387.html is that the L2 spatial prefetcher only concerns loads from memory to L2 and L3 cache and thus does not cause destructive interference, based on wording from the Intel optimization manual (this is the thread where gcc settled on 64 for the destructive interference size).

I went on and reproduced the tests at https://stackoverflow.com/a/72127222 on Intel i7-1185G7 (Tiger Lake). Starting from the benchmark provided there, I observed:
- No benefit of 128-alignment over 64-alignment if there is at most a thread per cache line. I also did not manage to reproduce the (negligible) effect on the counter `machine_clears.memory_order` mentioned in the answer.
- Moderate benefit of 128-alignment over 64-alignment when: 1) there are two threads competing on each cache line (4 threads total), 2) the operations are RMW (e.g. fetch_add). Moreover, the slowdown is around 20% (compare to 100% slowdown when all 4 threads try to access the same cache line). By running with hardware prefetcher turned off as a control, the slowdown disappears.

As a conclusion, the effect of the spatial prefetcher exists but is hard to reproduce, as it essentially only worsens pre-existing interference. It seems that the appropriate `Cache_line_size` on x86-64 is 64, because in order to avoid destructive interference from cache line pairs, it is enough to avoid destructive interference from single cache lines.

Reply:
Thank you for doing a deep dive on this. 

I've changed the `config.h` to 64b for amd64, riscv5 and non-native (this case is a little odd, in that I'm not sure if the value is actually being used outside tests).